### PR TITLE
Lane detection ROS2 node

### PR DESCRIPTION
**Pull request by leungjch**
Adds a ROS2 node for lane detection using `ultra-fast-lane-detection_v2` + TensorRT .

Source repo: https://github.com/iwatake2222/play_with_tensorrt/tree/master/pj_tensorrt_lane_ultra-fast-lane-detection_v2

Day demo:

https://github.com/WATonomous/wato_monorepo/assets/28817028/9fafa078-0e51-4cc6-9c6a-6d7787ab58fb

Night demo:

https://github.com/WATonomous/wato_monorepo/assets/28817028/4c0021a1-e390-4141-81f9-2feb707f2d19

### CODE REVIEW

File: docker/base/inject_opencv.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,8 @@
+ARG GENERIC_IMAGE
+
+########################## Install ROS2 Core ##########################
+FROM ${GENERIC_IMAGE} as wato_base
+
+RUN apt-get update
+RUN apt install -y libopencv-dev python3-opencv
+RUN export OpenCV_DIR=/usr/lib/x86_64-linux-gnu/cmake/opencv4/OpenCVConfig.cmake
```
Comment (Edwardius) [inline docker/base/inject_opencv.Dockerfile:8]: please do not touch base images. They should not be tampered with, especially when we are installing a very specific library like opencv. Just install this in the dependencies stage of the specific dockerfile you are building

**Review by Edwardius — DISMISSED** [review]

File: docker/base/inject_opencv.Dockerfile (hunk 1 above)
Comment (leungjch) [reply docker/base/inject_opencv.Dockerfile]: got it, removed now

File: modules/docker-compose.perception.yaml [hunk 2]
```diff
@@ -1,85 +1,87 @@
...
-      dockerfile: docker/perception/radar_object_detection/radar_object_detection.Dockerfile
-      cache_from: 
-        - "${PERCEPTION_RADAR_OBJECT_DETECTION_IMAGE:?}:${TAG}"
-        - "${PERCEPTION_RADAR_OBJECT_DETECTION_IMAGE:?}:main"
-      target: deploy
-    image: "${PERCEPTION_RADAR_OBJECT_DETECTION_IMAGE:?}:${TAG}"
-    security_opt:
-      - seccomp:${MONO_DIR}/modules/seccomp_profile.json 
-    command: /bin/bash -c "ros2 launch radar_object_detection radar_object_detection.launch.py"
+  # radar_object_detection:
```
Comment (leungjch) [inline modules/docker-compose.perception.yaml:4]: Uncomment before merging

File: docker/perception/lane_detection/lane_detection.Dockerfile [hunk 3]
```diff
@@ -43,13 +54,28 @@ RUN . /opt/ros/$ROS_DISTRO/setup.sh && \
 
 # Entrypoint will run before any CMD on launch. Sources ~/opt/<ROS_DISTRO>/setup.bash and ~/ament_ws/install/setup.bash
 COPY docker/wato_ros_entrypoint.sh ${AMENT_WS}/wato_ros_entrypoint.sh
-ENTRYPOINT ["./wato_ros_entrypoint.sh"]
+# ENTRYPOINT ["./wato_ros_entrypoint.sh]
 
 ################################ Prod ################################
-FROM build as deploy
+FROM ${RUNTIME_IMAGE} as deploy
```
Comment (Edwardius) [inline docker/perception/lane_detection/lane_detection.Dockerfile:60]: If a person runs this image as a dev container, can he/she rebuild the lane detection node inside it? eg. run colcon build while inside the container

Comment (Edwardius) [reply docker/perception/lane_detection/lane_detection.Dockerfile:60]: oh, nvm I already addressed this 😅

File: watod-config.sh [hunk 4]
```diff
@@ -12,15 +12,17 @@
...
 
 ############################## OPTIONAL CONFIGURATIONS ##############################
 ## Name to append to docker containers. DEFAULT = "<your_watcloud_username>"
-# COMPOSE_PROJECT_NAME=""
+COMPOSE_PROJECT_NAME="j89leung"
 
 ## Tag to use. Images are formatted as <IMAGE_NAME>:<TAG> with forward slashes replaced with dashes.
 ## DEFAULT = "<your_current_github_branch>"
-# TAG=""
+## TAG="lane-detection"
```
Comment (Edwardius) [inline watod-config.sh:25]: Change back

Comment (leungjch) [reply watod-config.sh]: done

**Comment by Edwardius** [comment]
oh dont mind tese comments for now, let me know when u need a review :)

**Comment by leungjch** [comment]
> oh dont mind tese comments for now, let me know when u need a review :)

sure feel free to re-review

File: src/wato_msgs/perception_msgs/lane_detection_msgs/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,25 @@
+cmake_minimum_required(VERSION 3.10)
```
Comment (leungjch) [inline src/wato_msgs/perception_msgs/lane_detection_msgs/CMakeLists.txt:1]: Added perception messages into a new directory `perception_msgs`

File: docker/perception/radar_object_detection/radar_object_detection.Dockerfile [hunk 6]
```diff
@@ -7,7 +7,7 @@ WORKDIR ${AMENT_WS}/src
 
 # Copy in source code 
 COPY src/perception/radar_object_detection radar_object_detection
-COPY src/wato_msgs/radar_msgs radar_msgs
+COPY src/wato_msgs/perception_msgs/radar_msgs radar_msgs
```
Comment (leungjch) [inline docker/perception/radar_object_detection/radar_object_detection.Dockerfile:10]: Grouping radar_msgs into perception_msgs folder

File: src/perception/lane_detection/CMakeLists.txt [hunk 7]
```diff
@@ -1,14 +1,64 @@
...
-# further dependencies manually.
-# find_package(<dependency> REQUIRED)
+find_package(rclcpp REQUIRED)
+find_package(std_msgs REQUIRED)
+find_package(sensor_msgs REQUIRED)
+find_package(geometry_msgs REQUIRED)
+find_package(lane_detection_msgs REQUIRED)
+find_package(cv_bridge REQUIRED)
+find_package(image_transport REQUIRED)
+find_package(CUDA REQUIRED)
```
Comment (Edwardius) [inline src/perception/lane_detection/CMakeLists.txt:45]: I think these can be combined...

File: src/perception/lane_detection/launch/eve.launch.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    )
+
+    return LaunchDescription(
+        [
+            Node(
+                package="lane_detection",
+                executable="lane_detection",
+                name="lane_detection_node",
+                parameters=[config],
+            )
```
Comment (Edwardius) [inline src/perception/lane_detection/launch/eve.launch.py:19]: Follow https://github.com/WATonomous/wato_monorepo/tree/aee5e7def1c1d6e694f84db05d8fe7210c3697e8/src/perception/camera_object_detection/launch format

File: src/perception/lane_detection/launch/nuscenes.launch.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+
+def generate_launch_description():
+    config = os.path.join(
+        get_package_share_directory("lane_detection"), "config", "nuscenes_config.yaml"
+    )
+
+    return LaunchDescription(
+        [
+            Node(
```
Comment (Edwardius) [inline src/perception/lane_detection/launch/nuscenes.launch.py:14]: follow https://github.com/WATonomous/wato_monorepo/tree/aee5e7def1c1d6e694f84db05d8fe7210c3697e8/src/perception/camera_object_detection/launch format

File: src/perception/lane_detection/package.xml [hunk 10]
```diff
@@ -4,13 +4,26 @@
   <name>lane_detection</name>
   <version>0.0.0</version>
   <description>TODO: Package description</description>
-  <maintainer email="e23zhou@uwaterloo.ca">bolty</maintainer>
+  <maintainer email="perception@watonomous.ca">perception</maintainer>
```
Comment (Edwardius) [inline src/perception/lane_detection/package.xml:7]: you can add ur own name :)

File: src/perception/lane_detection/package.xml [hunk 11]
```diff
@@ -4,13 +4,26 @@
...
   <license>TODO: License declaration</license>
 
   <buildtool_depend>ament_cmake</buildtool_depend>
 
+  <test_depend>ament_lint_auto</test_depend>
+  <test_depend>ament_lint_common</test_depend>
```
Comment (Edwardius) [inline src/perception/lane_detection/package.xml:13]: we arent using ament lint

File: src/perception/lane_detection/src/lane_detection_node.cpp [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+  rclcpp::Publisher<lane_detection_msgs::msg::LaneDetection>::SharedPtr
+      lane_detection_pub_;  // The lane detection publisher
+
+  size_t count_;
+
+  // ROS Parameters
+  std::string camera_topic;
+  std::string publish_vis_topic;
+  std::string publish_lanes_topic;
+  bool save_images;
```
Comment (Edwardius) [inline src/perception/lane_detection/src/lane_detection_node.cpp:156]: add _ at the end

File: src/perception/lane_detection/src/lane_detection_node.cpp [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+      std::filesystem::create_directories(save_dir);
+    }
+
+    RCLCPP_INFO(this->get_logger(), "Subscribing to camera topic: %s", camera_topic.c_str());
+
+    subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
+        camera_topic, 10,
+        std::bind(&LaneDetectionNode::image_callback, this, std::placeholders::_1));
+
+    image_pub_ = this->create_publisher<sensor_msgs::msg::Image>(publish_vis_topic, 10);
```
Comment (Edwardius) [inline src/perception/lane_detection/src/lane_detection_node.cpp:56]: Is this only meant for debugging? if so, only init and run this for debugging purposes

**Review by Edwardius — CHANGES_REQUESTED** [review]

File: src/perception/lane_detection/src/lane_detection_tensorrt
[the remark this answers is not in the document]
Comment (Gongsta) [reply src/perception/lane_detection/src/lane_detection_tensorrt:1]: @leungjch needs to add the `.gitmodules` file

File: src/perception/lane_detection/src/lane_detection_node.cpp (hunk 13 above)
Comment (Gongsta) [reply src/perception/lane_detection/src/lane_detection_node.cpp]: meant for visualization, similar to how we do it for camera object detection node

File: src/perception/lane_detection/src/lane_detection_node.cpp [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    {
+      cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
+    }
+    catch (cv_bridge::Exception &e)
+    {
+      RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
+      return;
+    }
+
+    cv::Mat image = cv_ptr->image.clone();
```
[the remark this answers is not in the document]
Comment (Gongsta) [reply src/perception/lane_detection/src/lane_detection_node.cpp:70]: @leungjch

File: src/perception/lane_detection/src/lane_detection_node.cpp
[the remark this answers is not in the document]
Comment (leungjch) [reply src/perception/lane_detection/src/lane_detection_node.cpp:1]: added now

File: src/perception/lane_detection/src/lane_detection_node.cpp [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+  for (size_t i = 0; i < raw_lane_list.size(); i++) {
+    lane_detection_msgs::msg::LaneLine line;
+    line.points.resize(raw_lane_list[i].size());
+    for (size_t j = 0; j < raw_lane_list[i].size() / 2; j++) {
+      line.points[j].x = raw_lane_list[i][j * 2 + 0];
+      line.points[j].y = raw_lane_list[i][j * 2 + 1];
+      line.points[j].z = 0;
+    }
+    lane_msg.lines.push_back(line);
+  }
```
Comment (Edwardius) [inline src/perception/lane_detection/src/lane_detection_node.cpp:108]: suggestion: move to separate function

File: src/perception/lane_detection/src/lane_detection_node.cpp [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+  if (publish_source_image_) {
+    lane_msg.source_img = *img_msg;
+  }
+
+  if (save_images_) {
+    std::string filename = save_dir_ + "/lane_detection_" + std::to_string(count_) + ".png";
+    cv::imwrite(filename, image);
+    RCLCPP_INFO(this->get_logger(), "Saved image to %s", filename.c_str());
+  }
```
Comment (Edwardius) [inline src/perception/lane_detection/src/lane_detection_node.cpp:118]: suggestion: move to separate function

File: src/perception/lane_detection/src/lane_detection_node.cpp [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    std::filesystem::create_directories(save_dir_);
+  }
+
+  RCLCPP_INFO(this->get_logger(), "Subscribing to camera topic: %s", camera_topic_.c_str());
+
+  subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
+      camera_topic_, 10,
+      std::bind(&LaneDetectionNode::image_callback, this, std::placeholders::_1));
+
+  image_pub_ = this->create_publisher<sensor_msgs::msg::Image>(publish_vis_topic_, 10);
```
Comment (Edwardius) [inline src/perception/lane_detection/src/lane_detection_node.cpp:48]: suggestion: Have this be a `debug` field in the config. It should only be initialized and ran when that field is set to true.

File: src/perception/lane_detection/src/lane_detection_node.cpp [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+  RCLCPP_INFO(this->get_logger(), "Total:               %9.3lf [msec]", time_all);
+  RCLCPP_INFO(this->get_logger(), "    Pre processing:  %9.3lf [msec]", result.time_pre_process);
+  RCLCPP_INFO(this->get_logger(), "    Inference:       %9.3lf [msec]", result.time_inference);
+  RCLCPP_INFO(this->get_logger(), "    Post processing: %9.3lf [msec]", result.time_post_process);
+  RCLCPP_INFO(this->get_logger(), "=== Finished frame ===");
+
+  // Convert the processed cv::Mat back to sensor_msgs::msg::Image and publish
+  sensor_msgs::msg::Image::SharedPtr img_msg =
+      cv_bridge::CvImage(msg->header, "bgr8", image).toImageMsg();
+  image_pub_->publish(*img_msg);
```
Comment (Edwardius) [inline src/perception/lane_detection/src/lane_detection_node.cpp:90]: ditto, only publish when `debug: true`

File: src/perception/lane_detection/src/lane_detection_node.cpp [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+  this->get_parameter("save_dir", save_dir_);
+  this->get_parameter("publish_source_image", publish_source_image_);
+
+  if (!std::filesystem::exists(save_dir_)) {
+    std::filesystem::create_directories(save_dir_);
+  }
+
+  RCLCPP_INFO(this->get_logger(), "Subscribing to camera topic: %s", camera_topic_.c_str());
+
+  subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
```
Comment (Edwardius) [inline src/perception/lane_detection/src/lane_detection_node.cpp:44]: This needs a roadmap in the future for handling multiple cameras.

Comment (Edwardius) [reply src/perception/lane_detection/src/lane_detection_node.cpp:44]: eg. how to store the subscribers, how to do inferencing in an optimal way, how to deal with overlapping inferences

File: watod-config.sh
Comment (Edwardius) [inline watod-config.sh:1]: undo this change

**Review by Edwardius — CHANGES_REQUESTED** [review]

File: src/perception/lane_detection/src/lane_detection_node.cpp (hunk 19 above)
Comment (leungjch) [reply src/perception/lane_detection/src/lane_detection_node.cpp:46]: this will be a future task that we'll have to think more about https://app.clickup.com/t/86dre2vka

for optimal inference we'll need to research how to do batched inference in tensorrt (see task desc)

we probably also need to ensure the camera input are synchronized, maybe we need an upstream node that subscribes [Camera1, Camera2, ..] , synchronizes their messages by timestamp, and publishes BatchedCameraImages , so all of our camera nodes can do batch processing without worrying about camera callbacks being triggered at different times

**Review by Edwardius — APPROVED** [review]
